pkgs/development/compilers/dotnet/patch-nupkgs.nix

```diff
@@ -31,34 +31,42 @@ writeShellScriptBin "patch-nupkgs" (
   ''
     set -euo pipefail
     shopt -s nullglob
-    isELF() {
-        local fn="$1"
-        local fd
-        local magic
-        exec {fd}< "$fn"
-        read -r -n 4 -u "$fd" magic
-        exec {fd}<&-
-        if [ "$magic" = $'\177ELF' ]; then return 0; else return 1; fi
-    }
     cd "$1"
   ''
   + lib.optionalString stdenv.hostPlatform.isLinux ''
+    isNonStaticExecutableELF() {
+      local magic
+      local e_type
+      local interpreter
+
+      if magic=$(od --address-radix=n --read-bytes=4 --format=c "$1") \
+        && [ "$magic" = " 177   E   L   F" ] \
+        && e_type=$(od --address-radix=n --skip-bytes=0x10 --read-bytes=2 --format=x2 "$1") \
+        && interpreter=$(${patchelf}/bin/patchelf --print-interpreter "$1") \
+        && [ "$e_type" != " 0002" -o -n "$interpreter" ]; then return 0; else return 1; fi
+    }
+
     for x in */* */*; do
       # .nupkg.metadata is written last, so we know the packages is complete
       [[ -d "$x" ]] && [[ -f "$x"/.nupkg.metadata ]] \
         && [[ ! -f "$x"/.nix-patched ]] || continue
       echo "Patching package $x"
       find "$x" -type f -print0 | while IFS= read -rd "" p; do
         if [[ "$p" != *.nix-patched ]] \
-          && isELF "$p" \
-          && ${patchelf}/bin/patchelf --print-interpreter "$p" &>/dev/null; then
+          && isNonStaticExecutableELF "$p"; then
           tmp="$p".$$.nix-patched
           # if this fails to copy then another process must have patched it
           cp --reflink=auto "$p" "$tmp" || continue
           echo "Patchelfing $p as $tmp"
-          ${patchelf}/bin/patchelf \
-            --set-interpreter "${stdenv.cc.bintools.dynamicLinker}" \
-            "$tmp" ||:
+
+          # Only patch the interpreter if it had one to begin with.
+          if interpreter=$(${patchelf}/bin/patchelf --print-interpreter "$p") \
+            && [ -n "$interpreter" ]; then
+            ${patchelf}/bin/patchelf \
+              --set-interpreter "${stdenv.cc.bintools.dynamicLinker}" \
+              "$tmp" ||:
+          fi
+
           # This makes sure that if the binary requires some specific runtime dependencies, it can find it.
           # This fixes dotnet-built binaries like crossgen2
           ${patchelf}/bin/patchelf \
```

Comment:
I believe this part will now be reached for shared libs, where it wasn't before (no interpreter).  Is that okay?

Reply:
It was before because even when there's no interpreter, `patchelf` still returns a zero exit code, which would've made it enter the `if` anyways.

Reply:
Nevermind, it seems it was a coincidence with the weird debug symbols that are in the other PR's package, reverting this change as well.